runtime/vm/jvminit.c

```diff
@@ -2843,8 +2843,41 @@ modifyDllLoadTable(J9JavaVM * vm, J9Pool* loadTable, J9VMInitArgs* j9vm_args)
 					return JNI_ERR;
 				}
 			}
+
+			/*
+			* On Linux on Z libj9jit dynamically loads libj9zlib as it is used for AOT method data compression
+			* which is currently only enabled on Z platform. We want to ensure that when the JVM loads libj9jit,
+			* libj9zlib is already loaded. See eclipse/openj9#8561 for more details.
+			*/
+#if (defined(S390) && defined(LINUX))
+			{
+			char zlibDll[EsMaxPath];
+			char *zlibDllDir = zlibDll;
+			UDATA expectedZlibPathLength = 0;
+			UDATA zlibDllLength = 0;
+			UDATA zlibFileHandle = 0;
+			
+			zlibDllLength = strlen(vm->j9libvmDirectory);
+			expectedZlibPathLength = zlibDllLength + (sizeof(DIR_SEPARATOR_STR) - 1) + strlen(J9_ZIP_DLL_NAME) + 1;
+			if (expectedZlibPathLength > EsMaxPath) {
+				zlibDllDir = j9mem_allocate_memory(expectedZlibPathLength, OMRMEM_CATEGORY_VM);
+				if (NULL == zlibDllDir) {
+					return JNI_ERR;
+				}
+			}
+			j9str_printf(PORTLIB, zlibDllDir, expectedZlibPathLength, "%s%s%s",
+					vm->j9libvmDirectory, DIR_SEPARATOR_STR, J9_ZIP_DLL_NAME);
+			zlibFileHandle = j9sl_open_shared_library(zlibDllDir, &(entry->descriptor), openFlags);
+			if (0 != zlibFileHandle) {
+				j9tty_printf(PORTLIB, "Error: Failed to open zlib DLL %s (%s)\n", zlibDllDir, j9error_last_error_message());
+				return JNI_ERR;
+			}
```

Comment:
Vaguely this code to do the free
```suggestion
			}
			if (zlibDllDir != zlibDll) {
				j9_mem_free_memory(zlibDllDir);
				zlibDllDir = NULL;
			}
```